frontend/src/pages/PairRoom/PairRoom.tsx

```diff
@@ -21,18 +23,35 @@ import * as S from './PairRoom.styles';
 
 const PairRoom = () => {
   const navigate = useNavigate();
+  const location = useLocation();
   const { accessCode } = useParams();
 
-  useEffect(() => {
-    const checkPairRoomExists = async () => {
-      if (!accessCode) navigate('/error');
-
-      const { exists } = await getPairRoomExists(accessCode || '');
+  const { addToast } = useToastStore();
 
-      if (!exists) navigate('/error');
+  useEffect(() => {
+    const checkAccessValid = async () => {
+      if (!location.state?.valid) {
+        navigate('/main');
+        addToast({ status: 'ERROR', message: '유효하지 않은 접근입니다. 올바른 경로로 접근해 주세요.' });
+        return;
+      }
+
+      if (!accessCode) {
+        navigate('/main');
+        addToast({ status: 'ERROR', message: '존재하지 않는 페어룸 코드입니다. 다시 확인해 주세요.' });
+        return;
+      }
+
+      const { exists } = await getPairRoomExists(accessCode);
+
+      if (!exists) {
+        navigate('/main');
+        addToast({ status: 'ERROR', message: '존재하지 않는 페어룸 코드입니다. 다시 확인해 주세요.' });
+        return;
+      }
```

Comment:
같은 메시지 2개면 조건문을 합쳐도 좋을 것 같아요!
```suggestion

      const { exists } = await getPairRoomExists(accessCode);
      
      if (!accessCode || !exists) {
        navigate('/main');
        addToast({ status: 'ERROR', message: '존재하지 않는 페어룸 코드입니다. 다시 확인해 주세요.' });
        return;
      }
```